### PR TITLE
[quant] update fused_obs_fake_quant op to accept output_fake_quant argument

### PR DESCRIPTION
**Pull request by supriyar**
Summary:
Add a new attribute to the FusedMovingAvgObsFakeQuantize that controls if the Fake Quant operation should be applied at the output of a particular layer. The motivation is to give the users additional control to control the numerics of the fake_quant operators during training. It defaults to always fake quant the output (True).

Note: We will still observer the tensors as before (only the fake_quant operation is controlled using this flag)

For example
```
input model
x -> fc1 -> fc2 -> non_quantizable_op -> fc3

After fake_quant
x -> fake_quant(x) -> fc1 -> fake_quant(fc1) -> fc2 -> fake_quant(fc2) -> non_quantizable_op -> fake_quant() -> fc3 -> fake_quantize(fc3)

With output_fake_quant disabled at the output of fc2 and fc3 (since their outputs are non-quantizable)
x -> fake_quant(x) -> fc1 -> fake_quant(fc1) -> fc2 -> non_quantizable_op -> fake_quant() -> fc3
```

Test Plan: ./buck-out/gen/caffe2/test/quantization_fx\#binary.par -r test_disable_output_fake_quant

Differential Revision: D31174526

### CODE REVIEW

File: test/quantization/fx/test_quantize_fx.py [hunk 1]
```diff
@@ -5507,6 +5507,81 @@ def forward(self, x):
...
+            qconfig = torch.quantization.get_default_qat_qconfig("fbgemm")
+            qconfig_dict = {
+                "": None,
+                "object_type": [
+                    (torch.nn.functional.linear, qconfig),
+                    (torch.nn.functional.relu, qconfig),
+                ],
+            }
+            prepared = torch.quantization.quantize_fx.prepare_qat_fx(model, qconfig_dict)
+            prepared.to(device)
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:5555]: nit: do we need to move again after prepare?

File: test/quantization/fx/test_quantize_fx.py [hunk 2]
```diff
@@ -5507,6 +5507,81 @@ def forward(self, x):
...
+                for user, _ in node.users.items():
+                    if user.op == 'call_function' and user.target == torch.nn.functional.linear:
+                        return False
+                return True
+
+
+            for node in prepared.graph.nodes:
+                if node.op == 'call_module' and torch.quantization.is_activation_post_process(modules[str(node.target)]):
+                    # get the users
+                    if check_all_users_not_linear(node):
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:5568]: is this what we are planning to do in practice?

File: test/quantization/fx/test_quantize_fx.py [hunk 3]
```diff
@@ -5507,6 +5507,81 @@ def forward(self, x):
...
+
+            def check_all_users_not_linear(node):
+                for user, _ in node.users.items():
+                    if user.op == 'call_function' and user.target == torch.nn.functional.linear:
+                        return False
+                return True
+
+
+            for node in prepared.graph.nodes:
+                if node.op == 'call_module' and torch.quantization.is_activation_post_process(modules[str(node.target)]):
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:5566]: nit: torch.ao.quantization

**Review by jerryzh168 — APPROVED** [review]

File: test/quantization/fx/test_quantize_fx.py (hunk 2 above)
Comment (supriyar) [reply test/quantization/fx/test_quantize_fx.py:5568]: this is one example (specific to linear) of how the output fake-quant can be disabled for a certain FakeQuantize module.

**Comment by janeyx99** [comment]
Hello, my team has been keeping track of reverts and the reasons for revert as a metric to approximate developer experience. Could you elaborate on why this PR was reverted for record-keeping? Thank you!
